### PR TITLE
feat(TimeLock): new TimeLock with upgrade compatible storage

### PR DESCRIPTION
**Pull request by re1ro**
* [x] `TimeLock`: base util contract, upgrade compatible

### CODE REVIEW

**Review by milapsheth — APPROVED** [review]
Looks good to me

File: contracts/interfaces/ITimeLock.sol [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+pragma solidity ^0.8.0;
+
+import { ITimeLock } from '../interfaces/ITimeLock.sol';
+
+interface ITimeLock {
+    error InvalidTimeLockHash();
+    error TimeLockAlreadyScheduled();
+    error TimeLockNotReady();
+
+    function MINIMUM_TIME_LOCK_DELAY() external view returns (uint256);
```
Comment (sammy1991106) [inline contracts/interfaces/ITimeLock.sol:12]:
```suggestion
    function minimum_time_lock_delay() external view returns (uint256);
```
Make `MINIMUM_TIME_LOCK_DELAY` internal instead and manually create a getter for it.

File: contracts/utils/TimeLock.sol [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        return _getTimeLockEta(hash);
+    }
+
+    function _scheduleTimeLock(bytes32 hash, uint256 eta) internal returns (uint256){
+        if (hash == 0) revert InvalidTimeLockHash();
+        if (_getTimeLockEta(hash) != 0) revert TimeLockAlreadyScheduled();
+
+        uint256 minimumEta = block.timestamp + MINIMUM_TIME_LOCK_DELAY;
+
+        if (eta < minimumEta) eta = minimumEta;
```
Comment (sammy1991106) [inline contracts/utils/TimeLock.sol:26]: why not error out instead of overwriting?

File: contracts/utils/TimeLock.sol [hunk 3]
```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import { ITimeLock } from '../interfaces/ITimeLock.sol';
+
+contract TimeLock is ITimeLock {
```
Comment (sammy1991106) [inline contracts/utils/TimeLock.sol:7]: Need to add unit tests for the `TimeLock` contract.

File: contracts/utils/TimeLock.sol [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    function _getTimeLockEta(bytes32 hash) private view returns (uint256 eta) {
+        bytes32 key = keccak256(abi.encodePacked(PREFIX_TIME_LOCK, hash));
+
+        assembly {
+            eta := sload(key)
+        }
+    }
+
+    function _setTimeLockEta(bytes32 hash, uint256 eta) private {
+        if (hash == 0) revert InvalidTimeLockHash();
```
Comment (deanamiel) [inline contracts/utils/TimeLock.sol:58]: I don't think this line of code is reachable, _setTimeLockEta is private and is called 3 times by other functions in this contract all of which have prior checks that that hash is not zero

File: contracts/utils/TimeLock.sol (hunk 3 above)
Comment (deanamiel) [reply contracts/utils/TimeLock.sol:7]: Unit tests PR [here](https://github.com/axelarnetwork/axelar-gmp-sdk-solidity/pull/42)

File: contracts/utils/TimeLock.sol (hunk 4 above)
Comment (re1ro) [reply contracts/utils/TimeLock.sol]: Removed

File: test/TimeLock.js [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  it('should successfully schedule a timelock', async () => {
+    const num = 5;
+    const numParam = defaultAbiCoder.encode(['uint256'], [num]);
+    const numHash = keccak256(numParam);
+    const timeDelay = 13 * 60 * 60;
+
+    const block = await ethers.provider.getBlock('latest');
+    const eta = block.timestamp + timeDelay;
+
+    await timeLock.scheduleSetNum(numHash, eta);
```
Comment (sammy1991106) [inline test/TimeLock.js:47]:
```suggestion
    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
```

File: test/TimeLock.js [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    const num = 5;
+    const numParam = defaultAbiCoder.encode(['uint256'], [num]);
+    const numHash = keccak256(numParam);
+    const minTimeDelay = 12 * 60 * 60;
+    const timeDelay = 2 * 60 * 60;
+
+    const block = await ethers.provider.getBlock('latest');
+    const eta = block.timestamp + timeDelay;
+
+    await timeLock.scheduleSetNum(numHash, eta);
```
Comment (sammy1991106) [inline test/TimeLock.js:63]:
```suggestion
    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
```

File: test/TimeLock.js [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  it('should revert when scheduling the same timelock twice', async () => {
+    const num = 5;
+    const numParam = defaultAbiCoder.encode(['uint256'], [num]);
+    const numHash = keccak256(numParam);
+    const timeDelay = 13 * 60 * 60;
+
+    const block = await ethers.provider.getBlock('latest');
+    const eta = block.timestamp + timeDelay;
+
+    await timeLock.scheduleSetNum(numHash, eta);
```
Comment (sammy1991106) [inline test/TimeLock.js:82]:
```suggestion
    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
```

File: test/TimeLock.js [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+  it('should cancel a timelock', async () => {
+    const num = 5;
+    const numParam = defaultAbiCoder.encode(['uint256'], [num]);
+    const numHash = keccak256(numParam);
+    const timeDelay = 13 * 60 * 60;
+
+    const block = await ethers.provider.getBlock('latest');
+    const eta = block.timestamp + timeDelay;
+
+    await timeLock.scheduleSetNum(numHash, eta);
```
Comment (sammy1991106) [inline test/TimeLock.js:105]:
```suggestion
    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
```

File: test/TimeLock.js [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+    const etaTimeLock = await timeLock.getTimeLock(numHash);
+    expect(etaTimeLock).to.equal(eta);
+
+    await timeLock.cancelSetNum(numHash);
```
Comment (sammy1991106) [inline test/TimeLock.js:110]:
```suggestion
    await timeLock.cancelSetNum(numHash).then((tx) => tx.wait());
```

File: test/TimeLock.js [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+  it('should revert on executing a timelock if buffer period has not passed', async () => {
+    const num = 5;
+    const numParam = defaultAbiCoder.encode(['uint256'], [num]);
+    const numHash = keccak256(numParam);
+    const timeDelay = 13 * 60 * 60;
+
+    const block = await ethers.provider.getBlock('latest');
+    const eta = block.timestamp + timeDelay;
+
+    await timeLock.scheduleSetNum(numHash, eta);
```
Comment (sammy1991106) [inline test/TimeLock.js:145]:
```suggestion
    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
```

File: test/TimeLock.js [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+  it('should execute timelock if buffer period has passed', async () => {
+    const num = 5;
+    const numParam = defaultAbiCoder.encode(['uint256'], [num]);
+    const numHash = keccak256(numParam);
+    const timeDelay = 13 * 60 * 60;
+
+    const block = await ethers.provider.getBlock('latest');
+    const eta = block.timestamp + timeDelay;
+
+    await timeLock.scheduleSetNum(numHash, eta);
```
Comment (sammy1991106) [inline test/TimeLock.js:165]:
```suggestion
    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
```

**Review by sammy1991106 — APPROVED** [review]

File: contracts/test/TimeLockTest.sol [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    event NumUpdated(uint256 newNum);
+
+    // solhint-disable-next-line no-empty-blocks
+    constructor(uint256 _minimumTimeDelay) TimeLock(_minimumTimeDelay) {}
+
+    function getNum() external view returns (uint256) {
+        return num;
+    }
+
+    function scheduleSetNum(bytes32 _hash, uint256 _eta) external {
```
Comment (milapsheth) [inline contracts/test/TimeLockTest.sol:19]: Similar to our governance contract, the `hash` should be derived from the params on-chain, `num` in this case, to better emulate the usage

File: test/TimeLock.js [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+
+    const block = await ethers.provider.getBlock('latest');
+    const eta = block.timestamp + timeDelay;
+
+    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
+
+    const etaTimeLock = await timeLock.getTimeLock(numHash);
+    expect(etaTimeLock).to.equal(eta);
+
+    await network.provider.send('evm_increaseTime', [timeDelay]);
```
Comment (milapsheth) [inline test/TimeLock.js:170]: Make this test hardhat agnostic, so we can run it on a live network. Use a smaller delay, 30s, for e.g. and if network is not hardhat, then sleep for that time. Can do this in a separate PR

File: test/TimeLock.js [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+    const eta = block.timestamp + timeDelay;
+
+    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
+
+    const minEta = block.timestamp + minTimeDelay;
+
+    const etaTimeLock = await timeLock.getTimeLock(numHash);
+
+    // allow for small block timestamp disparity
+    expect(Math.abs(etaTimeLock - minEta)).to.be.lessThanOrEqual(5);
```
Comment (milapsheth) [inline test/TimeLock.js:70]: why is this needed? such workarounds usually mean there's an incorrect expectation by the test. Since the block timestamp when scheduling is `1` more than when the latest block was fetched, you just need to add 1 to minEta. Better yet, to make this network agnostic, get the block number from the tx receipt when scheduling the eta

File: test/TimeLock.js [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+
+    await timeLock.scheduleSetNum(numHash, eta).then((tx) => tx.wait());
+
+    const etaTimeLock = await timeLock.getTimeLock(numHash);
+    expect(etaTimeLock).to.equal(eta);
+
+    await timeLock.cancelSetNum(numHash).then((tx) => tx.wait());
+
+    const cancelledEta = await timeLock.getTimeLock(numHash);
+    expect(cancelledEta).to.equal(0);
```
Comment (milapsheth) [inline test/TimeLock.js:113]: also check if the timelock can be scheduled again after cancellation

File: contracts/utils/TimeLock.sol [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+/**
+ * @title TimeLock
+ * @author Kiryl Yermakou
+ * @dev A contract that enables function execution after a certain time has passed.
+ * Implements the ITimeLock interface.
+ */
+contract TimeLock is ITimeLock {
+    bytes32 internal constant PREFIX_TIME_LOCK = keccak256('time-lock');
+
+    uint256 internal immutable MINIMUM_TIME_LOCK_DELAY;
```
Comment (milapsheth) [inline contracts/utils/TimeLock.sol:16]: Use camel case for naming. I think the convention is that only constants (not immutables) should use capital snake case.

File: contracts/utils/TimeLock.sol [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+    constructor(uint256 minimumTimeDelay) {
+        MINIMUM_TIME_LOCK_DELAY = minimumTimeDelay;
+    }
+
+
+    /**
+     * @notice Returns a minimum time delay at which the TimeLock may be scheduled.
+     * @return uint Unix time diff for a minimum scheduling delay
+     */
+    function minimumTimeLockDelay() external view override returns (uint256) {
```
Comment (milapsheth) [inline contracts/utils/TimeLock.sol:31]: query this in the test for coverage

File: contracts/utils/TimeLock.sol [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+        _setTimeLockEta(hash, 0);
+    }
+
+    /**
+     * @notice Executes an existing timelock and sets its eta back to zero.
+     * @dev To execute, the timelock must currently exist and the required time delay
+     * must have passed.
+     * @param hash The hash of the timelock to execute
+     */
+    function _executeTimeLock(bytes32 hash) internal {
```
Comment (Foivos) [inline contracts/utils/TimeLock.sol:79]: This name is confusing since there is no execution taking place here. Maybe change name to _checkAndCancelTimelock?

File: contracts/test/TimeLockTest.sol (hunk 12 above)
Comment (deanamiel) [reply contracts/test/TimeLockTest.sol:19]: Makes sense, I changed it so the hash is calculated on-chain and made a separate function to test the invalid hash edge case.

File: test/TimeLock.js (hunk 14 above)
Comment (deanamiel) [reply test/TimeLock.js:70]: Changed the test to get the block timestamp from the tx receipt when scheduling the eta.

File: test/TimeLock.js (hunk 15 above)
Comment (deanamiel) [reply test/TimeLock.js:113]: Added test case

File: contracts/utils/TimeLock.sol (hunk 17 above)
Comment (deanamiel) [reply contracts/utils/TimeLock.sol:29]: added test

File: contracts/utils/TimeLock.sol (hunk 18 above)
Comment (re1ro) [reply contracts/utils/TimeLock.sol]: renamed to `_finalizeTimeLock`

File: contracts/utils/TimeLock.sol [hunk 19]
```diff
@@ -12,22 +12,22 @@ import { ITimeLock } from '../interfaces/ITimeLock.sol';
 contract TimeLock is ITimeLock {
     bytes32 internal constant PREFIX_TIME_LOCK = keccak256('time-lock');
 
-    uint256 internal immutable MINIMUM_TIME_LOCK_DELAY;
+    uint256 internal immutable MinimumTimeLockDelay;
```
Comment (milapsheth) [inline contracts/utils/TimeLock.sol:15]:
```suggestion
    uint256 internal immutable _minimumTimeLockDelay;
```
perhaps

File: contracts/utils/TimeLock.sol [hunk 20]
```diff
@@ -12,22 +12,22 @@ import { ITimeLock } from '../interfaces/ITimeLock.sol';
...
 
     /**
      * @notice The constructor for the TimeLock.
      * @param minimumTimeDelay The minimum time delay (in secs) that must pass for the TimeLock to be executed
      */
     constructor(uint256 minimumTimeDelay) {
-        MINIMUM_TIME_LOCK_DELAY = minimumTimeDelay;
+        MinimumTimeLockDelay = minimumTimeDelay;
```
Comment (milapsheth) [inline contracts/utils/TimeLock.sol:22]:
```suggestion
        _minimumTimeLockDelay = minimumTimeDelay;
```

File: contracts/utils/TimeLock.sol [hunk 21]
```diff
@@ -12,22 +12,22 @@ import { ITimeLock } from '../interfaces/ITimeLock.sol';
...
     }
 
     /**
      * @notice Returns a minimum time delay at which the TimeLock may be scheduled.
      * @return uint Minimum scheduling delay time (in secs) from the current block timestamp
      */
     function minimumTimeLockDelay() external view override returns (uint256) {
-        return MINIMUM_TIME_LOCK_DELAY;
+        return MinimumTimeLockDelay;
```
Comment (milapsheth) [inline contracts/utils/TimeLock.sol:30]:
```suggestion
        return _minimumTimeLockDelay;
```

File: contracts/utils/TimeLock.sol [hunk 22]
```diff
@@ -51,7 +51,7 @@ contract TimeLock is ITimeLock {
         if (hash == 0) revert InvalidTimeLockHash();
         if (_getTimeLockEta(hash) != 0) revert TimeLockAlreadyScheduled();
 
-        uint256 minimumEta = block.timestamp + MINIMUM_TIME_LOCK_DELAY;
+        uint256 minimumEta = block.timestamp + MinimumTimeLockDelay;
```
Comment (milapsheth) [inline contracts/utils/TimeLock.sol:54]:
```suggestion
        uint256 minimumEta = block.timestamp + _minimumTimeLockDelay;
```

File: contracts/utils/TimeLock.sol [hunk 23]
```diff
@@ -76,7 +76,7 @@ contract TimeLock is ITimeLock {
      * must have passed.
```
Comment (milapsheth) [inline contracts/utils/TimeLock.sol:76]: update docs to use finalize vs execute

**Review by milapsheth — APPROVED** [review]
